### PR TITLE
api, models: finish adding type hints, make `ConfigurationKey` a proper enum (Bug 1759890)

### PR DESCRIPTION
**Pull request by cgsheeh**
Finish adding type hints to all modules in `api/` and `models/`.

Type hints for `put` are updated from `int` to `str` to reflect
the correct state in `swagger.yml`.

Add a type alias for configuration values that is an alias to
`bool | int | str`. Also raise at the end of the function to
avoid an accidental `None` return in some edge cases.
Convert `ConfigurationKey` to a proper `Enum` type, add
type hints for it and convert usages of it's underlying `str`
to use the `.value` attribute.

Add `flask_sqlalchemy.BaseQuery` as the return type for
model queries.

### CODE REVIEW

File: landoapi/models/configuration.py [hunk 1]
```diff
@@ -4,12 +4,19 @@
...
+    Union,
+)
+
 from landoapi.cache import cache
 from landoapi.models.base import Base
 from landoapi.storage import db
 
 logger = logging.getLogger(__name__)
 
+ConfigurationValue = Union[bool, int, str]
```
Comment (zzzeid) [inline landoapi/models/configuration.py:18]: Something about the name that makes me think this is a value and not a type. Perhaps we could just append "Type" to this.

```suggestion
ConfigurationValueType = Union[bool, int, str]
```

File: landoapi/models/transplant.py [hunk 2]
```diff
@@ -87,23 +90,23 @@ def update_from_transplant(self, landed, error="", result=""):
...
+    def head_revision(self) -> str:
         """Human-readable representation of the branch head's Phabricator revision ID."""
         assert (
             self.revision_order
         ), "head_revision should never be called without setting self.revision_order!"
         return "D" + self.revision_order[-1]
 
     @classmethod
-    def revisions_query(cls, revisions):
+    def revisions_query(cls, revisions: list[str]) -> Query:
```
Comment (zzzeid) [inline landoapi/models/transplant.py:105]: I think this is technically `flask_sqlalchemy.BaseQuery`.

**Review by zzzeid — CHANGES_REQUESTED** [review]

**Comment by cgsheeh** [comment]
Added some changes around `ConfigurationKey` to this patch FYI.

**Review by zzzeid — CHANGES_REQUESTED** [review]
I think the enum change no longer matches with the title of this commit, it should go in its own commit or the title should be changed.

**Comment by cgsheeh** [comment]
@zzzeid I updated the title, let me know if this is better or if you have a suggestion otherwise.

**Review by zzzeid — APPROVED** [review]
I would shorten it to something like:
`api, models: add more type hints, use enum (bug 1759890)`

And put the rest in the message.

**Pull request merged**